File: scripts/weather.ts
/**
 * Enrichit les courses TriRace avec des données météo historiques via Open-Meteo Archive API.
 *
 * - Remplit avg_temp_celsius et avg_wind_kmh pour chaque course ayant latitude + longitude + date
 * - Si la date est dans le futur, calcule la moyenne sur les 3 dernières années à la même date
 * - avg_water_temp_celsius est laissé NULL (pas d'API fiable gratuite)
 * - Traitement par batch de 10 en parallèle (Open-Meteo est généreux)
 *
 * Usage : npx tsx scripts/weather.ts
 * Requis : SUPABASE_SERVICE_KEY dans .env.local (pour bypasser la RLS)
 */

import { createClient } from '@supabase/supabase-js';
import { readFileSync } from 'fs';
import { resolve } from 'path';

// Load .env.local manually (dotenv not required)
try {
  const envPath = resolve(process.cwd(), '.env.local');
  const lines = readFileSync(envPath, 'utf-8').split('\n');
  for (const line of lines) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;
    const eq = trimmed.indexOf('=');
    if (eq === -1) continue;
    const key = trimmed.slice(0, eq).trim();
    const value = trimmed.slice(eq + 1).trim();
    if (key && !(key in process.env)) process.env[key] = value;
  }
} catch {}

const SUPABASE_URL = process.env.NEXT_PUBLIC_SUPABASE_URL || '';
const SUPABASE_KEY = process.env.SUPABASE_SERVICE_KEY || process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY || '';

if (!SUPABASE_URL || !SUPABASE_KEY) {
  console.error('❌ SUPABASE_URL et SUPABASE_SERVICE_KEY requis dans .env.local');
  process.exit(1);
}

const supabase = createClient(SUPABASE_URL, SUPABASE_KEY);

// --- Open-Meteo ---

const OPEN_METEO_BASE = 'https://archive-api.open-meteo.com/v1/archive';

interface OpenMeteoResponse {
  daily: {
    time: string[];
    temperature_2m_mean: (number | null)[];
    windspeed_10m_max: (number | null)[];
  };
}

/**
 * Fetches temperature_2m_mean and windspeed_10m_max for a given date and location.
 * Returns null values if the API call fails or data is missing.
 */
async function fetchWeatherForDate(
  lat: number,
  lng: number,
  date: string // YYYY-MM-DD
): Promise<{ temp: number | null; wind: number | null }> {
  const url =
    `${OPEN_METEO_BASE}` +
    `?latitude=${lat}` +
    `&longitude=${lng}` +
    `&start_date=${date}` +
    `&end_date=${date}` +
    `&daily=temperature_2m_mean,windspeed_10m_max` +
    `&timezone=Europe%2FParis`;

  try {
    const res = await fetch(url, {
      headers: { 'User-Agent': 'TriRace/1.0 (trirace.fr)' },
    });

    if (!res.ok) return { temp: null, wind: null };

    const data: OpenMeteoResponse = await res.json();
    const temp = data.daily?.temperature_2m_mean?.[0] ?? null;
    const wind = data.daily?.windspeed_10m_max?.[0] ?? null;

    return { temp, wind };
  } catch {
    return { temp: null, wind: null };
  }
}

/**
 * Returns the date string YYYY-MM-DD shifted by the given number of years (negative = past).
 */
function shiftYear(date: string, years: number): string {
  const d = new Date(date);
  d.setFullYear(d.getFullYear() + years);
  return d.toISOString().slice(0, 10);
}

/**
 * For a past date, fetches weather directly.
 * For a future date, averages over the 3 preceding years at the same calendar date.
 *
 * Open-Meteo archive only covers up to ~5 days ago, so any date after
 * TODAY_MINUS_5 is considered "future" and handled with historical averaging.
 */
async function getWeather(
  lat: number,
  lng: number,
  date: string // YYYY-MM-DD
): Promise<{ temp: number | null; wind: number | null }> {
  const today = new Date();
  const cutoff = new Date(today);
  cutoff.setDate(cutoff.getDate() - 5); // archive lag ~5 days

  const raceDate = new Date(date);

  if (raceDate <= cutoff) {
    // Past date: fetch directly
    return fetchWeatherForDate(lat, lng, date);
  }

  // Future date: average over the last 3 years at the same calendar date
  const years = [-1, -2, -3];
  const results = await Promise.all(
    years.map((offset) => fetchWeatherForDate(lat, lng, shiftYear(date, offset)))
  );

  const validTemps = results.map((r) => r.temp).filter((v): v is number => v !== null);
  const validWinds = results.map((r) => r.wind).filter((v): v is number => v !== null);

  const avg = (arr: number[]) =>
    arr.length > 0 ? Math.round((arr.reduce((a, b) => a + b, 0) / arr.length) * 10) / 10 : null;

  return {
    temp: avg(validTemps),
    wind: avg(validWinds),
  };
}

// --- Concurrency helper ---

async function runBatches<T>(
  items: T[],
  fn: (item: T, index: number) => Promise<void>,
  concurrency: number
): Promise<void> {
  for (let i = 0; i < items.length; i += concurrency) {
    await Promise.all(items.slice(i, i + concurrency).map((item, j) => fn(item, i + j)));
  }
}

// --- Main ---

interface RaceRow {
  id: number;
  slug: string;
  date: string;
  latitude: number;
  longitude: number;
}

async function main() {
  // Fetch courses with geo + date but without weather data yet
  const { data: races, error } = await supabase
    .from('races')
    .select('id, slug, date, latitude, longitude')
    .not('latitude', 'is', null)
    .not('longitude', 'is', null)
    .not('date', 'is', null)
    .is('avg_temp_celsius', null)
    .limit(5000);

  if (error || !races) {
    console.error('❌ Erreur Supabase:', error?.message);
    process.exit(1);
  }

  if (races.length === 0) {
    console.log('✓ Toutes les courses éligibles ont déjà des données météo.');
    return;
  }

  const total = races.length;
  console.log(`🌤  ${total} courses à enrichir avec la météo`);
  console.log(`📡 Source : Open-Meteo Archive API (gratuite, sans clé)`);
  console.log(`⚡ Concurrence : 10 requêtes en parallèle`);
  console.log('');

  let done = 0;
  let failed = 0;

  const CONCURRENCY = 10;

  await runBatches(
    races as RaceRow[],
    async (race) => {
      const { temp, wind } = await getWeather(race.latitude, race.longitude, race.date);

      if (temp === null && wind === null) {
        failed++;
        process.stdout.write(
          `\r  ✓ ${done} enrichies  ✗ ${failed} échouées  (${done + failed}/${total} traitées)   `
        );
        return;
      }

      const updatePayload: Record<string, number | null> = {
        avg_temp_celsius: temp,
        avg_wind_kmh: wind,
      };

      const { error: updateError } = await supabase
        .from('races')
        .update(updatePayload)
        .eq('id', race.id);

      if (updateError) {
        failed++;
      } else {
        done++;
      }

      process.stdout.write(
        `\r  ✓ ${done} enrichies  ✗ ${failed} échouées  (${done + failed}/${total} traitées)   `
      );
    },
    CONCURRENCY
  );

  console.log(`\n\n✓ Terminé : ${done} courses enrichies, ${failed} sans résultat.`);
}

main().catch((e) => {
  console.error('❌', e);
  process.exit(1);
});

File: scripts/enrichers/weather.py
"""
Enrichissement météo via Open-Meteo API (gratuit, pas de clé requise).
https://open-meteo.com/en/docs/historical-weather-api

Pour chaque course avec lat/lon et date, on récupère les moyennes historiques
(5 dernières années) : température haute, température basse, vent.
"""
import time
import requests
from datetime import date, timedelta
from typing import Optional


OPEN_METEO_URL = "https://archive-api.open-meteo.com/v1/archive"

# Nombre d'années d'historique à moyenner
HISTORY_YEARS = 5


def enrich_weather(supabase_client) -> int:
    """
    Enrichit toutes les courses sans météo qui ont lat/lon et une date.
    Retourne le nombre de courses enrichies.
    """
    print("[WEATHER] Enrichissement météo des courses...")

    # Récupérer les courses sans météo mais avec coordonnées
    result = (
        supabase_client.table("races")
        .select("id, slug, date, latitude, longitude, avg_temp_high_celsius")
        .is_("avg_temp_high_celsius", "null")
        .not_.is_("latitude", "null")
        .not_.is_("longitude", "null")
        .not_.is_("date", "null")
        .execute()
    )

    races = result.data or []
    print(f"[WEATHER] {len(races)} courses à enrichir")

    enriched = 0
    for race in races:
        try:
            weather = fetch_historical_weather(
                lat=race["latitude"],
                lon=race["longitude"],
                race_date=race["date"],
            )

            if weather:
                supabase_client.table("races").update(weather).eq("id", race["id"]).execute()
                enriched += 1
                print(
                    f"  ✓ {race['slug']} → "
                    f"high {weather.get('avg_temp_high_celsius')}°C / "
                    f"low {weather.get('avg_temp_low_celsius')}°C, "
                    f"vent {weather.get('avg_wind_kmh')} km/h"
                )

        except Exception as e:
            print(f"  ✗ {race['slug']} : {e}")

        time.sleep(0.3)  # Open-Meteo : max ~10 req/s, on est très conservateurs

    print(f"[WEATHER] {enriched}/{len(races)} courses enrichies")
    return enriched


def fetch_historical_weather(lat: float, lon: float, race_date: str) -> Optional[dict]:
    """
    Calcule la météo moyenne sur les 5 dernières années à la même date.
    Retourne un dict avec avg_temp_high_celsius, avg_temp_low_celsius, avg_wind_kmh.
    """
    try:
        race_dt = date.fromisoformat(race_date)
    except (ValueError, TypeError):
        return None

    today = date.today()
    temps_high = []
    temps_low = []
    winds = []

    for years_back in range(1, HISTORY_YEARS + 1):
        target_year = race_dt.year - years_back
        if target_year < 1940:
            break

        # Fenêtre de ±3 jours autour de la date de course
        window_start = date(target_year, race_dt.month, min(race_dt.day, 28)) - timedelta(days=3)
        window_end = window_start + timedelta(days=6)

        # Ne pas aller dans le futur
        if window_start > today:
            continue

        params = {
            "latitude": lat,
            "longitude": lon,
            "start_date": window_start.isoformat(),
            "end_date": min(window_end, today).isoformat(),
            "daily": "temperature_2m_max,temperature_2m_min,wind_speed_10m_max",
            "wind_speed_unit": "kmh",
            "timezone": "auto",
        }

        try:
            resp = requests.get(OPEN_METEO_URL, params=params, timeout=15)
            resp.raise_for_status()
            data = resp.json()

            daily = data.get("daily", {})
            high_list = [t for t in (daily.get("temperature_2m_max") or []) if t is not None]
            low_list  = [t for t in (daily.get("temperature_2m_min") or []) if t is not None]
            wind_list = [w for w in (daily.get("wind_speed_10m_max") or []) if w is not None]

            if high_list:
                temps_high.append(sum(high_list) / len(high_list))
            if low_list:
                temps_low.append(sum(low_list) / len(low_list))
            if wind_list:
                winds.append(sum(wind_list) / len(wind_list))

        except Exception:
            continue

        time.sleep(0.2)

    if not temps_high:
        return None

    result = {
        "avg_temp_high_celsius": round(sum(temps_high) / len(temps_high), 1),
        "avg_temp_low_celsius":  round(sum(temps_low) / len(temps_low), 1) if temps_low else None,
        "avg_wind_kmh":          round(sum(winds) / len(winds), 1) if winds else None,
    }

    return result
